### PR TITLE
[spinel] add an API to get the radio spinel metrics

### PR DESCRIPTION
**Pull request by zhanglongxia**
When the RCP is not stable, it causes the Thread stack crashed.
To better know the stability of the RCP, this PR adds an API to
return the radio spinel metrics.

### CODE REVIEW

File: src/lib/spinel/radio_spinel_metrics.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * This structure represents the radio spinel metrics.
+ *
+ */
+typedef struct otRadioSpinelMetrics
+{
+    uint32_t mRcpTimeoutCount;         ///< The number of RCP timeouts.
+    uint32_t mRcpUnexpectedResetCount; ///< The number of RCP unexcepted resets.
+    uint32_t mRcpRestorationCount;     ///< The number of RCP restorations.
+    uint32_t mSpinelParseErrorCount;   ///< The number of spinel frame parse errors.
```
Comment (jinran-google) [inline src/lib/spinel/radio_spinel_metrics.h:51]: Is it possible to overflow uint32 if there are many errors?

Comment (zhanglongxia) [reply src/lib/spinel/radio_spinel_metrics.h:51]: These errors are critical errors, they will cause the RCP restoration. The RCP restoration triggers the hardware reset of the RCP chip. The RCP chip takes a long time (250ms in our product) to start up, and the Host has to wait for the RCP to start up. This causes the counter won't increase rapidly. So I think`uint32` is enough. Thoughts?

Comment (jinran-google) [reply src/lib/spinel/radio_spinel_metrics.h:51]: Considering the extreme case where it triggers 4 times every second, it takes 34 years to overflow. uint32 should be good enough.

**Review by jinran-google — APPROVED** [review]
LGTM :+1:

**Review by abtink — APPROVED** [review]
LGTM. 👍

**Review by bukepo — APPROVED** [review]

File: src/lib/spinel/radio_spinel.hpp [hunk 2]
```diff
@@ -877,6 +878,14 @@ template <typename InterfaceType, typename ProcessContextType> class RadioSpinel
      */
     otError SendReset(uint8_t aResetType);
 
+    /**
+     * This method returns the radio Spinel metrics.
+     *
+     * @returns The radio Spinel metrics.
+     *
+     */
+    const otRadioSpinelMetrics *GetRadioSpinelMetrics(void) { return &mRadioSpinelMetrics; }
```
Comment (jwhui) [inline src/lib/spinel/radio_spinel.hpp:887]: Declare as `const`?
```suggestion
    const otRadioSpinelMetrics *GetRadioSpinelMetrics(void) const { return &mRadioSpinelMetrics; }
```

Comment (zhanglongxia) [reply src/lib/spinel/radio_spinel.hpp]: Updated

**Review by jwhui — APPROVED** [review]

**Pull request merged**